src/core/qgscoordinatereferencesystem.h

```diff
@@ -46,38 +46,189 @@ class QgsCoordinateReferenceSystem;
+ * available in our database of CRS). Whenever a new CRS definition is seen, it will
+ * be added to the local databse (in user's home directory, see QgsApplication::qgisUserDbFilePath()).
+ * QGIS also features a GUI for management of local custom CRS definitions.
+ *
+ * There are therefore two databases: one for shipped CRS definitions and one for custom CRS definitions.
+ * Custom CRS have internal IDs (accessible with srsid()) greater or equal to \ref USER_CRS_START_ID.
+ * The local CRS databases should never be accessed directly with SQLite functions, instead
+ * you should use QgsCoordinateReferenceSystem API for CRS lookups and for managements of custom CRS.
+ *
+ * Validation
+ * ==========
+ *
+ * In some cases (most prominently when loading a map layer), QGIS will try to ensure
+ * that the given map layer CRS is valid using validate() call. If not, a custom
+ * validation function will be called - such function may for example show a GUI
+ * for manual CRS selection. The validation function is configured using setCustomSrsValidation().
+ * If validation fails or no validation function is set, the default CRS is assigned
+ * (WGS84). QGIS application registers its validation function that will act according
+ * to user's settings (either show CRS selector dialog or use project/custom CRS).
+ *
+ * Object Construction and Copying  {#crs_construct_and_copy}
+ * ===============================
+ *
+ * The easiest way of creating CRS instances is to use QgsCoordinateReferenceSystem(const QString&)
+ * constructor that automatically recognizes definition format from the given string.
+ *
+ * Creation of CRS object involves some queries in a local SQLite database, which may
+ * be potentially expensive. It is therefore recommended to use QgsCRSCache methods
+ * that return possibly cached CRS objects instead of constructing new instances that
+ * involve the lookup overhead.
+ *
+ * Since QGIS 2.16 QgsCoordinateReferenceSystem objects are implicitly shared.
+ *
+ * The following table summarizes equivalents for non-cached and cached CRS lookup:
+ *
+ * Definition  | Non-cached            | Cached
+ * ----------- | --------------------- | ------------------------------
+ * Auth + Code | createFromOgcWmsCrs() | QgsCRSCache::crsByOgcWmsCrs()
+ * PROJ.4      | createFromProj4()     | QgsCRSCache::crsByProj4()
+ * WKT         | createFromWkt()       | QgsCRSCache::crsByWkt()
+ *
+ * Caveats
+ * =======
+ *
+ * There are two different flavours of WKT: one is defined by OGC, the other is the standard
+ * used by ESRI. They look very similar, but they are not the same. QGIS is able to consume
+ * both flavours.
+ *
+ * \see QgsCoordinateTransform, QgsCRSCache
+ *
  */
 class CORE_EXPORT QgsCoordinateReferenceSystem
 {
   public:
 
+    //! Enumeration of types of IDs accepted in createFromId() method
     enum CrsType
     {
-      InternalCrsId,
-      PostgisCrsId,
-      EpsgCrsId     // deprecated
```

Comment:
Eventually I would like to get rid of the whole enum, but that's a different story :)